pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,587 @@
+
+```slice
+struct   // The struct keyword
+\struct  // An unscoped identifier
+\foo     // An unscoped identifier
+
+::\struct::\module // A globally scoped identifier
+```
+
+### Integer literals
+
+Slice supports three forms of integer literals:
+- A decimal literal for base-10 numbers. These consist of one or more digits between `0` and `9`.
+- A hexadecimal literal for base-16 numbers. These start with the prefix `0x`, followed by one or more digits between `0` and `9` or letters between `a` and `f`. Hexadecimal literals are not case-sensitive.
+- A binary literal for base-2 numbers. These start with the prefix `0b`, followed by one or more digits that are either `0` or `1`.
+
+Additionally, underscores can appear between any two characters of an integer literal.
+These underscores carry no semantic meaning and are discarded by the parser.
+
+Examples of valid integer literals:
+
+```slice
+123
+0xFF
+0b1010
+
+335_445_996
+0x_ab_cd_ef  // Same as 0xabcded
+0b0_______1  // Same as 0b01
+
+0x_0b1101  // hexadecimal literal of value 725249
+
+// All have a value of 0
+0
+0_00_0
+0b000
+0x0
+```
+
+{% callout type="note" %}
+Note that `-` signs are counted as separate tokens, and not as part of an integer literal.
+{% /callout %}
+
+The Slice grammar has 2 separate rules for unsigned and (possibly) signed integers.
+Slice does not support using a `+` sign to indicate an integer is positive, since all integers are positive by default.
+
+```ebnf {% showTitle=false %}
+Integer
+    : integer_literal
+    ;
+
+SignedInteger
+    : Integer
+    | "-" Integer
+    ;
+```
+
+### String literals
+
+String literals consist of any number of UTF8 characters, wrapped in a pair of double quotes.
+Any characters within a string literal are treated as opaque text by the parser.
+
+Slice supports escaping special characters within string literals.
+This allows you to include unprintable characters or literal double quote characters in your string.
+Characters can be escaped by prefixing them with a backslash.
+To write a literal backslash, you must escape it with another backslash: `\\`.
+
+The parser resolves any escaped characters, removing the `\` escape character where present.
+Apart from this, no additional processing is performed on string literals.
+
+Examples of valid string literals:
+
+```slice
+""
+"hello world"
+"the variable \"foo\" doesn't exist"
+"a backslash: '\\'"
+```
+
+String literals can only appear as arguments to an [attribute].
+
+### Attributes
+
+An attribute consists of a directive, optionally followed by attribute arguments.
+
+A directive is syntactically equivalent to a [scoped identifier][identifier], but instead of specifying namespaces, the presence of scopes limits when the attribute is applicable.
+When no scopes are specified, the attribute is always applied, but when scopes are present, it is applied conditionally.
+For example, `deprecated` is always applied, but `cs::type` will only be applied when compiling to C#.
```

Comment:
This is somehow up to compiler implementation, from the language point of view is just a scope for the attribute.

Reply:
I just removed the last 2 sentences completely, since we couldn't find a wording that ticked all the boxes.
Now we just say:
```
A directive is syntactically equivalent to a [scoped identifier][identifier], but instead of specifying namespaces,
the presence of scopes limits when the attribute is applied.
```